crates/sparrow-compiler/src/table_content.rs

```diff
@@ -0,0 +1,63 @@
+use tonic::codegen::BoxStream;
+
+use crate::Error;
+
+pub struct TableContent {
+    schema: SchemaRef,
+    version: usize,
+    merged: RecordBatch,
+    updates: tokio::sync::broadcast::Sender<(usize, RecordBatch)>,
+}
+
+impl TableContent {
+    pub fn new(schema: SchemaRef) -> Self {
+        let (updates, _) = tokio::sync::broadcast::channel(1);
+        let merged = RecordBatch::new_empty(schema.clone());
+        mSelf {
+            schema,
+            version: 0,
+            merged,
+            updates,
+        }
+    }
+
+    pub fn add_batch(&mut self, batch: RecordBatch) -> error_stack::Result<(), Error> {
+        if batch.num_rows() == 0 {
+            return Ok(());
+        }
+
+        let merged = if self.merged.num_rows() == 0 {
+            batch
+        } else {
+            homogeneous_merge(&self.schema, vec![self.merged.clone(), batch])
+                .into_report()
+                .change_context(Error::Internal("add_batch"))?
+        };
+
+        self.version += 1;
+        self.updates
+            .send((self.version, merged.clone()))
+            .into_report()
+            .change_context(Error::Internal("add_batch"))?;
+        self.merged = merged;
+        Ok(())
+    }
+
+    pub fn stream(&self) -> BoxStream<RecordBatch> {
+        async_stream::stream! {
+            let mut version = self.version;
+            let mut recv = self.updates.subscribe();
+
+            tracing::info!("Starting subscriber with version {version}");
+            yield self.merged;
+            while let Ok((recv_version, batch)) = recevier.recv().await {
+                tracing::
+                if version < recv_version {
```

Comment:
Logic here being to avoid race conditions where multiple callers `add_batch` and send updates out of order? Does there need to be a lock around the `add_batch` method? i.e. is it possible we have 
```
Self { batch: A} 
add_batch(B)
add_batch(C)
merge(A + B)
merge(A + C)
self.updates.send(A+B)
self.updates.send(A+C)
self.merge = A+B
self.merge=A+C
```

Reply:
In this PR -- this is fine, because `add_batch` has `&mut self` which means it is the only reference. In the next PR, I had to fix that by introducing the lock.